runtime/near-vm-runner/src/errors.rs

```diff
@@ -58,33 +64,21 @@ impl IntoVMError for wasmer_runtime::error::ResolveError {
 
 impl IntoVMError for wasmer_runtime::error::RuntimeError {
     fn into_vm_error(self) -> VMError {
-        use wasmer_runtime::error::RuntimeError;
-        match &self {
-            RuntimeError::Trap { msg } => {
-                VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: msg.to_string() })
-            }
-            RuntimeError::Error { data } => {
-                if let Some(err) = data.downcast_ref::<VMLogicError>() {
-                    match err {
-                        VMLogicError::HostError(h) => {
-                            VMError::FunctionCallError(FunctionCallError::HostError(h.clone()))
-                        }
-                        VMLogicError::ExternalError(s) => VMError::ExternalError(s.clone()),
-                        VMLogicError::InconsistentStateError(e) => {
-                            VMError::InconsistentStateError(e.clone())
-                        }
-                    }
-                } else {
-                    eprintln!(
-                        "Bad error case! Output is non-deterministic {:?} {:?}",
-                        data.type_id(),
-                        self.to_string()
-                    );
-                    VMError::FunctionCallError(FunctionCallError::WasmTrap {
-                        msg: "unknown".to_string(),
-                    })
+        let data = &*self.0;
+
+        if let Some(err) = data.downcast_ref::<VMLogicError>() {
+            match err {
+                VMLogicError::HostError(h) => {
+                    VMError::FunctionCallError(FunctionCallError::HostError(h.clone()))
+                }
+                VMLogicError::ExternalError(s) => VMError::ExternalError(s.clone()),
+                VMLogicError::InconsistentStateError(e) => {
+                    VMError::InconsistentStateError(e.clone())
                 }
             }
+        } else {
+            // Reusing https://github.com/wasmerio/wasmer/blob/3999728e39016517c655aaa08e7f7e0e6c42cc58/lib/runtime-core/src/error.rs#L190
+            VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: format!("{}", self) })
```

Comment:
Is it deterministic? The reason we used unknown is to avoid being non-deterministic traps.

Reply:
Traps are specific to the spec, so they cannot be non-deterministic (https://webassembly.org/docs/security/).. 

Another question if Runtime can return internal errors also. 

So our current strategy is: if we can't identify the type of the error we assume that it's an internal Wasmer error, correct?

Reply:
I've asked in the Wasmer chat
https://spectrum.chat/wasmer/runtime/runtimeerror-cause-internal-error~53d40c84-dc1c-426f-9d83-ab6db8c96c23

Reply:
It's quite sketchy to rely on the implementation here: https://docs.rs/wasmer-runtime-core/0.16.2/src/wasmer_runtime_core/error.rs.html#190

It can be one of 3 types (str, String, and Exception) and also `unknown`.  We need to explicitly confirm with Wasmer that the error is deterministic from their end and independent from the hardware.

Reply:
From my understanding:
- RuntimeError can be either a type from wasmer or a user-specified type
- Display implementation handles wasmer type and String in case user-defined type is a string for convenience
- Because of a bug in at least earlier wasmer versions, the error in some cases was actually a string, and in addition to that a non-deterministic one (contained some address values for example)
- In some cases (don't remember which, could have been with metering feature) the error was neither a string nor any of the types

As long as ExceptionCode is the only possible case Display should be fine, but would be better to assert that

Reply:
An answer from Wasmer:

> So it looks like we return an ExceptionCode for some types of traps. It seems that internal issues are returned as strings and are not uniquely identifiable as coming from the Wasmer internals.
I think this is something that we may need to update, I'll ping the team about it!

Reply:
Looks like the most safe we can do is to terminate the node if the error is not `VMLogicError` or `ExceptionCode`, correct?
@evgenykuzyakov @mikhailOK @nearmax

Reply:
according to the quote we can't do that yet, would the assert hold on our existing tests?

Reply:
looks like tests are passing

Reply:
https://github.com/nearprotocol/nearcore/issues/2316#issuecomment-608497650